hmpps-prisoner-search/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/attributesearch/api/Matchers.kt

```diff
@@ -0,0 +1,46 @@
+package uk.gov.justice.digital.hmpps.prisonersearch.search.services.attributesearch.api
+
+import uk.gov.justice.digital.hmpps.prisonersearch.search.services.attributesearch.AttributeSearchException
+
+data class Matchers(
+  val joinType: JoinType,
+  val stringMatchers: List<StringMatcher>? = null,
+  val booleanMatchers: List<BooleanMatcher>? = null,
+  val intMatchers: List<IntMatcher>? = null,
+  val dateMatchers: List<DateMatcher>? = null,
+  val dateTimeMatchers: List<DateTimeMatcher>? = null,
+  val children: List<Matchers>? = null,
+) {
+  fun validate() {
+    if (typeMatchers().isEmpty() && children.isNullOrEmpty()) {
+      throw AttributeSearchException("Matchers must not be empty")
+    }
+  }
+
+  fun typeMatchers(): List<TypeMatcher<*>> =
+    listOfNotNull(stringMatchers, booleanMatchers, intMatchers, dateMatchers, dateTimeMatchers)
+      .flatten()
+}
+
+fun List<Matchers>.getAllMatchers(): List<Matchers> {
```

Comment:
how about using `fold` instead of the `forEach` i.e. 
```kotlin
fun List<Matchers>.getAllMatchers(): List<Matchers> =
  fold(mutableListOf()) { allMatchers, matcher ->
    allMatchers.add(matcher)
    matcher.children?.also { children -> allMatchers.addAll(children.getAllMatchers()) }
    allMatchers
  }
```